File: src/main/Java/kafka/avro/SimpleAvroProducer.java
package kafka.avro;

import com.twitter.bijection.Injection;
import com.twitter.bijection.avro.GenericAvroCodecs;
import org.apache.avro.Schema;
import org.apache.avro.generic.GenericData;
import org.apache.avro.generic.GenericRecord;
import org.apache.kafka.clients.producer.KafkaProducer;
import org.apache.kafka.clients.producer.ProducerRecord;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.Properties;

/**
 * Inspired by:
 * https://github.com/aseigneurin/kafka-sandbox
 * http://aseigneurin.github.io/2016/03/04/kafka-spark-avro-producing-and-consuming-avro-messages.html
 *
 */
public class SimpleAvroProducer {

    public static void main(String[] args) throws InterruptedException, IOException {
        Properties props = new Properties();
        props.put("bootstrap.servers", "localhost:9092");
        props.put("key.serializer", "org.apache.kafka.common.serialization.StringSerializer");
        props.put("value.serializer", "org.apache.kafka.common.serialization.ByteArraySerializer");

        String schemaFile = new String(Files.readAllBytes(Paths.get("./src/main/Java/kafka/avro/record.avsc")));

        Schema.Parser parser = new Schema.Parser();
        Schema schema = parser.parse(schemaFile);
        Injection<GenericRecord, byte[]> recordInjection = GenericAvroCodecs.toBinary(schema);

        KafkaProducer<String, byte[]> producer = new KafkaProducer<>(props);
        for (int i = 0; i < 1000; i++) {
            GenericData.Record avroRecord = new GenericData.Record(schema);
            avroRecord.put("str1", "Str 1-" + i);
            avroRecord.put("str2", "Str 2-" + i);
            avroRecord.put("int1", i);

            System.out.println("Sending record: " + avroRecord.toString());

            byte[] bytes = recordInjection.apply(avroRecord);

            ProducerRecord<String, byte[]> record = new ProducerRecord<>("avro-topic", bytes);
            producer.send(record);

            Thread.sleep(250);
        }

        producer.close();
    }
}

File: src/main/scala/alpakka/kafka/avro/SimpleAvroProducer.scala
package alpakka.kafka.avro

import com.sksamuel.avro4s.*
import org.apache.kafka.clients.producer.{KafkaProducer, ProducerRecord}
import org.apache.kafka.common.serialization.StringSerializer

import java.util.Properties

/**
  * Not pekko streams related
  *
  * Prerequisite:
  * Run [[alpakka.env.KafkaServerEmbedded]]
  */
object SimpleAvroProducer {
  def main(args: Array[String]): Unit = {
    new Application();
    ()
  }

  private class Application {
    val props = new Properties()
    props.put("bootstrap.servers", "localhost:29092")
    props.put("key.serializer", classOf[StringSerializer].getName)
    props.put("value.serializer", classOf[AvroSerializer].getName)

    val producer = new KafkaProducer[String, AvroRecord](props)

    try for (i <- 0 until 100) {
      val avroRecord = AvroRecord(s"Str 1-$i", s"Str 2-$i", i)
      println(s"Sending record: $avroRecord")

      val record = new ProducerRecord[String, AvroRecord]("avro-topic", avroRecord)
      producer.send(record)

      Thread.sleep(100)
    } finally producer.close()
  }
}

class AvroSerializer extends org.apache.kafka.common.serialization.Serializer[AvroRecord] {
  override def serialize(topic: String, data: AvroRecord): Array[Byte] = {
    val baos = new java.io.ByteArrayOutputStream()
    val avroOutputStream = AvroOutputStream.binary[AvroRecord].to(baos).build()
    avroOutputStream.write(data)
    avroOutputStream.close()
    baos.toByteArray
  }
}
